Beth Haubert

bethanyhaubert opened the pull request.

bethanyhaubert commented:
Database submission:
https://docs.google.com/spreadsheets/d/1C7xkzUFX_77UjQJDuWJdtVecksAXMsd6iGxL5_gpAk4/edit?usp=sharing

sumeetjain commented:
Consider making a table for "schools" and "conferences". A competitor definitely belongs to a school, but does a competitor also belong to a conference? Or would it make more sense for a _school_ to belong to a conference?

Re: Awards, your design is 100% valid–but allow me to suggest an alternative:

| event_id | first_place | second_place | third_place |
| --- | --- | --- | --- |
| 2 | 4 | 2 | 1 |
| 5 | 1 | 4 | 3 |

Another alternative:

| event_id | competitor_id | rank |
| --- | --- | --- |
| 2 | 4 | 1 |
| 2 | 2 | 2 |
| 2 | 1 | 3 |
| 5 | 1 | 1 |
| 5 | 4 | 2 |
| 5 | 4 | 3 |

I'll be frank: I'm not 100% sure which of these options is the best. But I wanted to show you, and you can pick any you like. If you're content with the one you have, that'll work.

bethanyhaubert commented:
DB update:
https://docs.google.com/spreadsheets/d/1C7xkzUFX_77UjQJDuWJdtVecksAXMsd6iGxL5_gpAk4/edit#gid=0

sumeetjain commented:
Make a 'conferences' table. In it, you'd have each conference.

Then a 'schools' table, where each school is listed.

bethanyhaubert commented:
https://docs.google.com/spreadsheets/d/1C7xkzUFX_77UjQJDuWJdtVecksAXMsd6iGxL5_gpAk4/edit#gid=1444586550

sumeetjain commented:
Nice! But you lost something in this process that you had before: Each contestant's time for their performance in an event. You'll need a table to track that.

bethanyhaubert commented:
https://docs.google.com/spreadsheets/d/1C7xkzUFX_77UjQJDuWJdtVecksAXMsd6iGxL5_gpAk4/edit#gid=1025594443

sumeetjain commented:
:+1:

bethanyhaubert commented:
finished all of the competitor actions that I could think of right now. I'm sure i'll be adding things later after you give us more direction, but was hoping for a thumbs up to continue with my other active record objects.

sumeetjain commented:
NICE. Yes, :+1:

bethanyhaubert commented:
I have working 'list', 'create', 'view', 'edit' pages/controllers for all of my models. Moving on to part 2 & 3

bethanyhaubert commented:
I want to add the name of the school next to each competitor, but i'm not sure how to do that in an embedded ruby loop. 

```
<h3> List of Competitors: </h3>
    <% @list_competitors.each do |c| %>
      <p><a href = "/profile/<%= c.id %>"> <%= c.first_name %> <%= c.last_name %> </a></p>
    <% end %>
```

it would be great if i could to `c.school_id.name`, but i don't know how to do it since 'c' isn't a controller variable and I don't think it's possible to make it one in my controller. At a standstill.

bethanyhaubert commented:
i think i'm going to just add schools/conference to competitor profiles instead for now

sumeetjain commented:
> i think i'm going to just add schools/conference to competitor profiles instead for now

Good decision. We'll go over how to connect model objects with other models later this week though.

bethanyhaubert commented:
I'm trying to get the school name to show up in my competitor_profile. Here's my controller:

```
MyApp.get "/profile/:competitor_id" do
  @competitor = Competitor.find(params[:competitor_id])
  x = @competitor.school_id
  @school = School.find({id: x})  <------ one way I've tried
  @school = School.find({id: params[x]}) <------ another way I've tried
  @school = School.find(x) <------ another way I've tried
  erb :"competitors/profile"
end
```

Here's the code in my profile view:

`<h3> <%= @competitor.first_name %> <%= @competitor.last_name %> - <%= @school.name %></h3>
`

Here are the errors:

 @school = School.find({id: params[x]})
 @school = School.find({id: x}) 

ActiveRecord::RecordNotFound at /profile/2
Couldn't find School with 'id'={:id=>nil}
file: core.rb location: find line: 155


@school = School.find(x):

ActiveRecord::RecordNotFound at /profile/2
Couldn't find School with 'id'=
file: core.rb location: find line: 155

I looked up finder methods for ActiveRecord and it seems like the School.find(x) should work, so i did binding.pry and it looks like 'x' isn't storing the value of '@competitor.school_id' even though 'school_id' exists in my schema for the competitors database. Need help moving forward.

bethanyhaubert commented:
I want to list all of the competitors who are competing in each event on the 'event_info' page
and I also want to list all of the events each competitor is competing in on their 'profile' page.

Currently, the database that holds the info about who is competing in which event is only in the results database. I feel like the database configuration could be better, but should I keep moving forward? or should i go back and add too another database or create a new one called 'entries' that consists of 'event_id', 'competitor_id'?

See current DBs:

```
DB.define_table("events")
DB.define_column("events", "name", "string")
DB.define_column("events", "first_place", "integer")
DB.define_column("events", "second_place", "integer")
DB.define_column("events", "third_place", "integer")

DB.define_table("competitors")
DB.define_column("competitors", "first_name", "string")
DB.define_column("competitors", "last_name", "string")
DB.define_column("competitors", "school_id", "integer")

DB.define_table("schools")
DB.define_column("schools", "name", "string")
DB.define_column("schools", "conference_id", "integer")

DB.define_table("conferences")
DB.define_column("conferences", "name", "string")

DB.define_table("results")
DB.define_column("results", "event_id", "integer")
DB.define_column("results", "competitor_id", "integer")
DB.define_column("results", "time", "string")
```

Possible new DB:

```
DB.define_table("entries")
DB.define_column("entries", "event_id", "integer")
DB.define_column("entries", "competitor_id", "integer")
```

sumeetjain commented:
Both would be valid, I think. With database design, it can be really hard to "get it right" the first time. I suggest moving on. If _I_ were building this, I would probably just use the `results` table as you are.

bethanyhaubert commented:
I can't get my erb to recognize the new method that I created for School class.

I created it in the model:

```
class School < ActiveRecord::Base

        #Take the School's conference_id and finds the name of the Conference object
        def conference_name
            x = self.conference_id
            y = Conference.find_by_id(x)
            y.name
        end
end
```

Here is the controller:

```
MyApp.get "/schools" do
    @list_schools = School.all
  erb :"schools/list_schools"
end

```

And I called the 'conference_name'  in my erb :"schools/list_schools"

```
    <% @list_schools.each do |s| %>
        <li>
            <a href = "/school_info/<%= s.id %>"> <%= s.name %> 
            </a>
            : 
            <a href = "/conference_info/<%= s.conference_id %>"> <%= s.conference_name %>
            </a>

```

```

Error:
NoMethodError at /schools
undefined method `name' for nil:NilClass
file: school.rb location: conference_name line: 7
```

line 7 is the "y.name" line

I don't know what else to do. I defined the method in the School model, then I called it in my erb file. so I can only imagine that the controller is lacking, but that could be wrong too.

sumeetjain commented:
What is the line of code in **school.rb** on line 7? Ruby says you're calling a method `name` in that line of code somewhere, and that you're calling it on `nil`.

If the above question/sentence don't show you the problem right away, that's okay. But you should respond here with the answer to the question at least.

bethanyhaubert commented:
fyi, in my db schema the name of the conference is called "name"

bethanyhaubert commented:
and I already mentioned that line 7 is the "y.name" line

bethanyhaubert commented:
i know i'm calling it on nil, i just can't figure out why

  y = Conference.find_by_id(x) should return a Conference object

sumeetjain commented:
Gotcha. What if a school has no `conference_id` value? Then it would return `nil`. Maybe you have a school with no conference set?

bethanyhaubert commented:
thanks, added if/else statement to method in model. now works

bethanyhaubert commented:
On each event details page I want to list each of the competitor that's currently entered in that event so i created a method to return

``` ruby
class Event < ActiveRecord::Base


    #this method returns a list of Competitor objects that match the competitor_id given in the array of 
        #Results which contain the matching event_id of the current Event object
    def list_of_results
        x = self.id
        #get the results that have this event_id
        list_of_results = Result.where({"event_id" => x})
        #with each of those results get the competitor_id
        if list_of_results == nil
            return nil
        else
            id_arr = []
            list_of_results.each do |result|
                id_arr << result.competitor_id
            end
        end

        competitors = []
        if id_arr == nil
            return nil

            #with the competitor.id find each Competitor object in the competitor table
            else id_arr.each do |c|
                competitors << Competitor.find_by_id(c)
                end
        end
        return competitors
    end
end
```

sumeetjain commented:
- If there are no results, I don't think `where` returns `nil`. It probably returns an empty array. So instead of `if list_of_results == nil`, you should try `if list_of_results.empty?`
- To avoid unanticipated Ruby syntax errors, split `else id_arr.each do |c|` into separate lines (`else` on its own line)

Everything else looks good to me. Are you getting any particular error?

bethanyhaubert commented:
``` ruby
class Event < ActiveRecord::Base


    #this method returns a list of Results object
    # that contain the event_id of the current Event
    def list_of_results
        x = self.id
        #get the results that have this event_id
        list_results = Result.where({"event_id" => x})
        #with each of those results get the competitor_id
        if list_results.empty?
            return nil
        else
            id_arr = []
            list_results.each do |result|
                id_arr << result.competitor_id
            end
        end

        competitors = []
        if id_arr.empty?
            return nil
        else
            #with the competitor.id find each Competitor object in the competitor table
             id_arr.each do |c|
                competitors << Competitor.find_by_id(c)
            end
        end

        return competitors
    end
```

Error:

```
NoMethodError at /event_info/1
undefined method `list_of_results' for nil:NilClass
file: event_info.erb location: block in singleton class line: 6
```

line 6:
`<% if @events.list_of_results == nil %>
`
I don't know what to do next. i changed the 'arr == nil's to arr.empty?. I also added to my erb so that if the method 'list_of_results' returns nil, it should show one comment, otherwise it should show the Competitor's names. 

```
    <ul>
    <% if @events.list_of_results == nil %>
        No competitors have been added.
        <% else %>
        <% @event.list_of_results.each do |c| %>
            <li>
                <a href="/profile/<%= c.id %>">
                    <%=c.first_name %> <%= c.last_name %>
                </a>
            </li>
            <% end %>
        <% end %>
    </ul>
```

what am i missing?

bethanyhaubert commented:
I also tried

```
 <ul>
    <% if @events.list_of_results.empty? %>
        No competitors have been added.
```

sumeetjain commented:
> I don't know what to do next. i changed the 'arr == nil's to arr.empty?.

That doesn't seem to be relevant here. You probably succeeded with that fix. Ruby is not saying that the `list_of_results` method is failing. It's saying it **cannot run** that method on `nil`. Which means that it's saying `@events` is `nil`.

> I also added to my erb so that if the method 'list_of_results' returns nil, it should show one comment, otherwise it should show the...

Again, if `@events` is `nil`, then checking whether `@events.list_of_results` returns `nil` or not is irrelevant. Ruby can't even run that method on `@events` at all.

The question is: Why is `@events` nil? What are your thoughts? Where is `@events` supposed to be defined? What might be causing it to be nil?

bethanyhaubert commented:
i don't know. Just to make sure, I checked my controller. I already have the controller variable @events in my controller for this view.

bethanyhaubert commented:
finished with Phase 3

bethanyhaubert commented:
I'm trying to implement a form on my event_info page that allows you to add a competitor to the 'results' database for the event_id of the page you're currently on. Here's my controller:

```
MyApp.get "/event_info/:event_id" do
  @event = Event.find(params[:event_id])
  @list_competitors = Competitor.all
  erb :"events/event_info"
end

MyApp.post "/add_entry" do
    x = Result.new
    x.event_id = self.id
    x.competitor_id = params[:competitor_id]
    x.save
  erb :"success"
end
```

I get that self.id isn't working in the entry form controller. I don't want to include an option to choose the event in the form, but i'm not sure how to automatically assign the event_id of the current page as x.event_id in the /add_entry controller

Error just for reference:
NoMethodError at /add_entry
undefined method `id' for #MyApp:0x007fe883acc7b8
file: event_controller.rb location: block in <top (required)> line: 25

line 25: x.event_id = self.id

sumeetjain commented:
Just like when you build the form to edit something, you need to dynamically set the form's action. It should be:

``` erb
<form action="/add_entry/<%= @event.id %>" method="post">
```

and update the controller action for it:

``` ruby
MyApp.post "/add_entry/:banana" do
```

That way, the controller action that processes the form will have access to `params[:banana]`.

bethanyhaubert commented:
thank you

bethanyhaubert commented:
I added a form that you can access directly from the event_info page that should allow the admin to add a time to the matching Result object column. But I'm getting an error that says that the 'time' method doesn't exist for 'Relation'. I thought that it had something to do with my 'where' method. like maybe this is just an error that is occurring because my database has ghosts of deleted competitors who have been added and then deleted from the results table.But i reset the database and I'm still getting the same error. This is my code in the controller:

```
MyApp.get "/add_result/time/:competitor_id/:event_id" do
    x = Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id])
    x.time = params[:time]
    x.save
  erb :"success"
end
```

NoMethodError at /add_result/time/:competitor_id/:event_id
undefined method `time=' for #<ActiveRecord::Relation []> Did you mean? timeout
file: delegation.rb location: method_missing line: 136

sumeetjain commented:
`.where` always returns a **collection**–even if there's only one row that the search results in.

So if you investigate what `x` is, in your code, you'll find that the result row _is_ there, but it's inside of the Array-like thing that ActiveRecord's `where` method returns. Try `x[0]` or `x.first` to pluck it out of the array. Then `.time` should work.

bethanyhaubert commented:
i tried '.first'
i got this error instead:
NoMethodError at /add_result/time/:competitor_id/:event_id
undefined method `time=' for nil:NilClass Did you mean? timeout

sumeetjain commented:
That sounds like `Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id])` is returning zero results. It's still like an Array–just an empty Array.

If you have `x = []` and do `x.first`, that will return `nil`.

So your code might be working just fine. The question is, why is `Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id])` returning no results? Presumably, there are no rows in the `results` table that match those constraints.

bethanyhaubert commented:
No, i checked. the array isn't empty

```
[2] pry(#<MyApp>)> Result.all
D, [2016-02-14T14:18:44.564499 #44951] DEBUG -- :   Result Load (0.7ms)  SELECT "results".* FROM "results"
=> [#<Result:0x007f943c8cae60 id: 1, created_at: 2016-02-14 20:05:49 UTC, updated_at: 2016-02-14 20:05:49 UTC, event_id: 1, competitor_id: 1, time: nil>,
 #<Result:0x007f943c872fd0 id: 2, created_at: 2016-02-14 20:05:57 UTC, updated_at: 2016-02-14 20:05:57 UTC, event_id: 1, competitor_id: 2, time: nil>,
 #<Result:0x007f943c872dc8 id: 3, created_at: 2016-02-14 20:06:03 UTC, updated_at: 2016-02-14 20:06:03 UTC, event_id: 1, competitor_id: 3, time: nil>]
[3] pry(#<MyApp>)>
```

bethanyhaubert commented:
but this is what `Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id]).first` is returning

```
[3] pry(#<MyApp>)> Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id]).first
D, [2016-02-14T14:21:01.236958 #44951] DEBUG -- :   
Result Load (0.8ms)  SELECT  "results".* FROM "results" WHERE "results"."competitor_id" = $1 AND "results"."event_id" = $2  ORDER BY "results"."id" ASC LIMIT 1 
 [["competitor_id", 0], ["event_id", 0]]
```

sumeetjain commented:
Note: The comment immediately above this is not what I'm replying to. I'm replying to the comment you typed that begins with `No, i checked. the array isn't empty`.

---

But the example you've just provided is not running the same code as your method is. You just showed me at `Result.all` is not empty. But what about `Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id])`?

You should add an interrupt like so:

``` ruby
MyApp.get "/add_result/time/:competitor_id/:event_id" do
        binding.pry
    result = Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id]).first
    result.time = params[:time]
    result.save
        erb :"success"
end
```

Then at that point, check what `params` has in it (to make sure that meets your expectations). And check what `Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id])` returns.

sumeetjain commented:
You need to check specifically what `Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id])` returns. Not `Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id]).first` -- you **know** that's `nil` already, because Ruby is telling you it's nil.

sumeetjain commented:
Once you check `Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id])`, you'll find that it's empty (I'm sure of it).

Then your question is, **why**? Verify that everything you assume is actually true. Do you have any results where the "competitor_id" is whatever `params[:competitor_id]` is? Etc.

bethanyhaubert commented:
no, see above. i _did_ check what it was returning

bethanyhaubert commented:
it returns the same thing

```
[4] pry(#<MyApp>)> 
Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id])

D, [2016-02-14T14:26:33.057648 #44951] DEBUG -- :   
Result Load (0.9ms)  SELECT "results".* 
FROM "results" WHERE "results"."competitor_id" = $1 AND "results"."event_id" = $2 
 [["competitor_id", 0], ["event_id", 0]]
```

WHERE "results"."competitor_id" = $1 AND "results"."event_id" = $2

sumeetjain commented:
There are **no results there**. Compare that to what happened when you typed `Result.all`:

```
=> [#<Result:0x007f943c8cae60 id: 1, created_at: 2016-02-14 20:05:49 UTC, updated_at: 2016-02-14 20:05:49 UTC, event_id: 1, competitor_id: 1, time: nil>,
 #<Result:0x007f943c872fd0 id: 2, created_at: 2016-02-14 20:05:57 UTC, updated_at: 2016-02-14 20:05:57 UTC, event_id: 1, competitor_id: 2, time: nil>,
 #<Result:0x007f943c872dc8 id: 3, created_at: 2016-02-14 20:06:03 UTC, updated_at: 2016-02-14 20:06:03 UTC, event_id: 1, competitor_id: 3, time: nil>]
```

Clearly, when there **are** results, Pry shows you them.

So again, my advice to you: http://cl.ly/f8k4/Screen%20Shot%202016-02-14%20at%202.28.16%20PM.png

bethanyhaubert commented:
I don't understand what you're saying. i did check to see what 'Result.where(competitor_id: params[:competitor_id], event_id: params[:event_id])' was. see above.

sumeetjain commented:
What do you expect `params` to be at that moment (in the `binding.pry`)? What is `params` actually?

sumeetjain commented:
What I'm trying to say is that **something you expect/intend** is not reality. So to debug, you have to figure out which of your assumptions is incorrect.

sumeetjain commented:
To do that, check _everything_.

bethanyhaubert commented:
i tried  doing params in pry, but i don't know how. i need different help

bethanyhaubert commented:
this isn't working

bethanyhaubert commented:
I'm having a hard time with the locking mechanism. and i feel like it is the reason i cannot find the syntax error. I've been over everything a dozen times
`event_controller.rb:55: syntax error, unexpected keyword_end, expecting end-of-input (SyntaxError)`

bethanyhaubert commented:
finished phase 5

bethanyhaubert commented:
oops, gotta fix my delete function

sumeetjain commented:
Are you seeking help/feedback?

bethanyhaubert commented:
I wrote this test

```
   def test_is_a_competitors_event_entry_empty
    assert_includes(@cat3.event_entry, nil)
  end
```

to test this method:

```
#this method finds the Result objects that match the Competitor object's id
#it then stores each Results' event_id in an Array
# then returns the Event that matches the event_id
        def event_entry
        list_results = Result.where({"competitor_id" => self.id})
        if list_results.empty?
            return nil
        else
            event_id_arr =[]
            list_results.each do |result|
                event_id_arr << result.event_id
            end
        end

        if event_id_arr.empty?
            return nil
        else
            return Event.where("id" => event_id_arr)
        end
    end
```

but i'm getting this error for that test:

```
 1) Failure:
CompetitorTest#test_is_a_competitors_event_entry_empty [/Users/ocs-guest1/Code/web-app__swimming-competition/tests/unit/competitor_test.rb:69]:
Expected nil (NilClass) to respond to #include?.
```

unsure what i should use in place of nil...

sumeetjain commented:
Your test looks valid:

``` ruby
assert_includes(@cat3.event_entry, nil)
```

Seems fine, because `assert_includes` should take parameter-1, which is a collection; and parameter-2, which is something that you're checking if parameter-1 includes. That is, it basically asks, "Does parameter-1 include parameter-2?".

Ruby seems to be telling you that `@cat3.event_entry`, which is your parameter-1, is `nil`. It's trying to check if parameter-1 includes parameter-2, but parameter-1 (`@cat3.event_entry`) is `nil`–and `nil` can't include anything.

So:
- [ ] Verify that `@cat3.event_entry` is `nil`, as we fear it might be. (Use `binding.pry` for this.)
- [ ] If it is `nil`, figure out why and fix that.

If `@cat3.event_entry` is not `nil`, let me know and I'll dig further.

bethanyhaubert commented:
when i run pry it says that @cat3 is nil. i don't know why. i defined it in setup

sumeetjain commented:
It's sort of hard to believe that `@cat3` is `nil`. If that were the case, I'd have expected an error when you tried to do `@cat3.event_entry` (because Ruby shouldn't allow you to run `event_entry` on `nil`).

I would be less surprised if `@cat3.event_entry` was `nil` though. Are you certain you're being accurate in your reporting here?

If so, can you come down and show this to me?

bethanyhaubert commented:
i was putting binding.pry inside my method. here's the thing though. @cat3.event_entry is _supposed_ to be nil, so why would i need to fix that?

sumeetjain commented:
Ah, okay. I understand now. The problem is that you're using the wrong assertion in your test.

You want to verify that `@cat3.event_entry` is `nil`. But the assertion you're using is checking that `@cat3.event_entry` **includes** `nil`.

Here's something that **includes** `nil`:

``` ruby
x = [nil, nil, nil]
```

But `nil` itself does not include `nil`. The assertion you want is:

``` ruby
assert_nil(@cat3.event_entry)
```

bethanyhaubert commented:
ng. thank you!

bethanyhaubert commented:
I'm also unsure of how to test delete methods:

```
    def delete_competitor_info
        Result.where("competitor_id" => self.id).delete_all
    end
```

Here's is my current test:

```
  def test_delete_competitor_info
    refute_includes(@cat2.delete_competitor_info, @result_3)
  end

```

and the failure:

```
 1) Failure:
  1) Failure:
CompetitorTest#test_delete_competitor_info [/Users/ocs-guest1/Code/web-app__swimming-competition/tests/unit/competitor_test.rb:73]:
Expected 1 (Fixnum) to respond to #include?.
```

sumeetjain commented:
That's a case where the _exercise_ and _verify_ steps are separate:

exercise: Run `@cat2.delete_competitor_info`

verify: Check that `@cat2` has zero competitors.

bethanyhaubert commented:
thank you
